lib/services/local/provisioning.go

```diff
@@ -189,11 +191,59 @@ func (s *ProvisioningService) GetTokens(ctx context.Context) ([]types.ProvisionT
 			services.WithRevision(item.Revision),
 		)
 		if err != nil {
-			return nil, trace.Wrap(err)
+			return nil, trace.Wrap(err, "unmarshaling token (key: %q)", item.Key)
 		}
 		tokens[i] = t
 	}
 	return tokens, nil
 }
 
+// ListProvisionTokens returns a paginated list of provision tokens.
+func (s *ProvisioningService) ListProvisionTokens(ctx context.Context, pageSize int, pageToken string, anyRoles types.SystemRoles, botName string) ([]types.ProvisionToken, string, error) {
+	// Bound page size (0 - 1_000)
+	if pageSize <= 0 || pageSize > int(defaults.MaxIterationLimit) {
+		pageSize = int(defaults.MaxIterationLimit)
+	}
+
+	prefix := backend.NewKey(tokensPrefix)
+	var out []types.ProvisionToken
+	for item, err := range s.Items(ctx, backend.ItemsParams{
+		StartKey: prefix.AppendKey(backend.KeyFromString(pageToken)),
+		EndKey:   backend.RangeEnd(prefix.ExactKey()),
+	}) {
+		if err != nil {
+			return nil, "", trace.Wrap(err)
+		}
+
+		t, err := services.UnmarshalProvisionToken(
+			item.Value,
+			services.WithExpires(item.Expires),
+			services.WithRevision(item.Revision),
+		)
+		if err != nil {
+			return nil, "", trace.Wrap(err, "unmarshaling token (key: %q)", item.Key)
+		}
+
+		if len(out) == pageSize {
+			nextKey := strings.TrimPrefix(
+				item.Key.TrimPrefix(backend.ExactKey(tokensPrefix)).String(),
+				string(backend.Separator),
+			)
+			return out, nextKey, nil
+		}
+
+		if len(anyRoles) > 0 && !t.GetRoles().IncludeAny(anyRoles...) {
```

Comment:
Let's put a note here that filtering logic needs to be kept in sync with the cache implementation. Or even better, perhaps extract a common predicate function?